wgsl/index.bs

```diff
@@ -6411,7 +6426,137 @@ Issue: (dneto) Default rounding mode is an implementation choice.  Is that what
+      a : f32;
+      b : u32;
+      c : f32;
+    };
+
+    [[group(0), binding(0)]]
+    var<storage> v : S;
+
+    fn foo() {
+      let x = v.b; // Does not access memory locations for v.a or v.c.
+    }
+  </xmp>
+</div>
+
+## Memory Model Reference ## {#memory-model-reference}
+
+Each module-scope variable in [SHORTNAME] forms a unique [memory model
+reference](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-references)
+for the lifetime of a given entry point.
+Each function-scope variable in [SHORTNAME] forms a unique [memory model
+reference](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-references)
+for the lifetime of the variable.
+
+## Scoped Operations ## {#scoped-operations}
+
+[[#atomic-builtin-functions|Atomic built-in functions]] map to [atomic
+operations](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-atomic-operation)
+whose memory
+[scope](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-scope)
+is:
+* `Workgroup` if the atomic pointer is in the [=storage classes/workgroup=]
+    storage class
+* `Device`/`QueueFamily` if the atomic pointer is in the [=storage
+    classes/storage=] storage class
+
+[[#sync-builtin-functions|Synchronization built-in functions]] map to control
+barriers whose execution
+[scope](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-scope)
+is `Workgroup` and whose memory
+[scope](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-scope)
+is:
+* `Workgroup` for `workgroupBarrier`
+* `Device`/`QueueFamily` for `storageBarrier`
+
+Note: If a `workgroupBarrier` and a `storageBarrier` are combined, the
+resulting memory scope is `Device`/`QueueFamily`.
+
+## Memory Semantics ## {#memory-semantics}
+
+All [[#atomic-builtin-functions|Atomic built-in functions]] use `Relaxed`
+[memory
+semantics](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-memory-semantics)
+and, thus, no storage class semantics.
+
+[[#sync-builtin-functions|workgroupBarrier]] uses `AcquireRelease` [memory
+semantics](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-memory-semantics)
+and `WorkgroupMemory` storage semantics.
+[[#sync-builtin-functions|storageBarrier]] uses `AcquireRelease` [memory
+semantics](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-memory-semantics)
+and `UniformMemory` storage semantics.
+
+Note: A combined `workgroupBarrier` and `storageBarrier` uses `AcquireRelease`
+ordering semantics and both `WorkgroupMemory` and `UniformMemory` memory
+semantics.
+
+Note: No atomic or synchronization built-in functions use `MakeAvailable` or
+`MakeVisible` semantics.
+
+## Private vs Non-Private ## {#private-vs-non-private}
```

Comment:
I had a very interesting discussion with @dneto0 in March 2019, on whether the mapping should make all read/write accesses be MakePointerVisible/MakePointerAvailable (the approach you've taken in this section, let's call it approach 1), or whether all barriers should have MakeAvailable/MakeVisible semantics (let's call it approach 2).
My (definitely fuzzy at this point) understanding is that both approaches work, and are probably equivalent in their results (but we don't have any proof at this stage). Approach 1 is what SPIRV-Tools does when converting "Coherent" annotations to the more modern Vulkan memory model, whereas approach 2 has the benefit of allowing some significant simplifications of the model. David also said at the time:
> Stepping back, the reason the Vulkan memory model has both per-op AV/VIS ops and also the bulk ones at synchronization points is that different workloads can perform better with one pattern vs. the other.

Considering how subtle this issue is, and how it is independent from the rest of the memory model, I intend to open a separate issue for it. It would at least ensure that all of the information on the topic is publicly on GitHub and no longer locked in email threads, and ideally we may get numbers on the perf impact of picking one approach over the other.
I'm fine with this PR landing with the current approach as waiting for a full discussion and consensus on this particular point may take a long while, I just want to make sure we don't forget about it.